contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,492 @@
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        uint128 positionLiquidity = position.liquidity;
+        require(positionLiquidity >= liquidity);
+
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+        // might be less gas optimal to remove this?
+        // uint256 feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        // uint256 feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+
+        position.tokensOwed0 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128,
+                positionLiquidity,
+                FixedPoint128.Q128
+            )
+        );
+        position.tokensOwed1 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside1LastX128 - position.feeGrowthInside1LastX128,
+                positionLiquidity,
+                FixedPoint128.Q128
+            )
+        );
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        // subtraction is safe because we checked positionLiquidity is gte liquidity
+        position.liquidity = uint128(positionLiquidity - liquidity);
+    }
+
+    // deploy ERC-20 contract
+    function beforeInitialize(address, IPoolManager.PoolKey calldata key, uint160) external override returns (bytes4) {
+        require(key.tickSpacing == 60, "Tick spacing must be default");
+
+        // deploy erc20 contract
+
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        // if we're at a new block, rebalance
+        if (block.number > blockNumber) {
+            blockNumber = block.number;
+            // retrieve all liquidity
+            uint128 hookLiquidity = poolManager.getLiquidity(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+            HookPosition storage position = poolToHookPosition[key.toId()];
+
+            position.liquidity = hookLiquidity;
+
+            IPoolManager.ModifyPositionParams memory params = IPoolManager.ModifyPositionParams({
+                tickLower: MIN_TICK,
+                tickUpper: MAX_TICK,
+                liquidityDelta: -int256(int128(hookLiquidity))
+            });
+
+            BalanceDelta balanceDelta = hookModifyPosition(key, params);
+
+            (uint160 sqrtPriceX96,,,,,) = poolManager.getSlot0(key.toId());
+
+            uint160 sqrtRatioAX96 = TickMath.getSqrtRatioAtTick(MIN_TICK);
```

Comment:
if this doesn't get used more than once, saves gas not to define it in memory (just stick it right in the params on line 438) having some gas tests will help see what's going on as you change things